plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java

```diff
@@ -1291,4 +1401,302 @@ private static TrinoException handleException(OrcDataSourceId dataSourceId, Exce
         }
         return new TrinoException(ICEBERG_CURSOR_ERROR, format("Failed to read ORC file: %s", dataSourceId), exception);
     }
+
+    private static RecordSet toRecordSet(ConnectorTransactionHandle sourceTransaction, IcebergSystemTable table, ConnectorSession session, TupleDomain<Integer> constraint)
+    {
+        return new RecordSet()
+        {
+            private final List<Type> types = table.getTableMetadata().getColumns().stream()
+                    .map(ColumnMetadata::getType)
+                    .collect(toImmutableList());
+
+            @Override
+            public List<Type> getColumnTypes()
+            {
+                return types;
+            }
+
+            @Override
+            public RecordCursor cursor()
+            {
+                return table.cursor(sourceTransaction, session, constraint);
+            }
+        };
+    }
+
+    private static class MappedPageSource
```

Comment:
Code copied from `trino-main`.

Any idea on how to avoid the code duplication?